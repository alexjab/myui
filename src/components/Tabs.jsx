import React from 'react'
import styled from 'styled-components'

import { light as lightTheme } from '../core/themes'

class BaseTab extends React.Component {
  selectTab = () => {
    const { index, selectTab } = this.props

    selectTab(index)
  }

  render() {
    const { children, className, label } = this.props

    return (
      <div className={className} onClick={this.selectTab}>
        {children}
        {label ? <Label>{label}</Label> : null}
      </div>
    )
  }
}

const Tab = styled(BaseTab)`
  padding-left: 15px;
  padding-right: 15px;
  padding-top: 13px;
  padding-bottom: 14px;

  color: ${({ isSelected, theme }) =>
    isSelected ? null : theme.colorGreyDark};
  border-bottom: ${({ isSelected, theme }) =>
    isSelected ? `3px solid ${theme.colorPrimary}` : 'none'};
  height: 1em;

  font-weight: 600;

  cursor: pointer;
  user-select: none;
`

Tab.defaultProps = {
  theme: lightTheme,
}

class BaseTabs extends React.Component {
  static defaultProps = {
    defaultTab: 0,
    tabs: [],
  }

  state = {
    selected: this.props.defaultTab,
  }

  selectTab = selected => {
    this.setState({ selected })

    const { onChangeSelected } = this.props
    if (onChangeSelected) {
      onChangeSelected(selected)
    }
  }

  render() {
    const { children, className, tabs, theme } = this.props
    const { selected } = this.state

    if (!tabs.length) {
      return null
    }

    return (
      <div className={className}>
        {tabs.map((tab, index) => (
          <Tab
            key={index}
            index={index}
            isSelected={selected === index}
            selectTab={this.selectTab}
            theme={theme}
            label={tab.label}
          >
            {tab.title}
          </Tab>
        ))}
      </div>
    )
  }
}

const Tabs = styled(BaseTabs)`
  font-family: ${props => props.theme.fontFamily};
  font-size: ${props => props.theme.fontSize};

  border-bottom: ${({ hasBorder, theme }) =>
    hasBorder ? `1px solid ${theme.colorGreyLighter}` : null};
  display: flex;
`

Tabs.defaultProps = {
  hasBorder: true,
  theme: lightTheme,
}

const Label = styled.span`
  color: ${({ theme }) => theme.colorGreyDarker};
  font-weight: 600;
  margin-left: 10px;
`

Label.defaultProps = {
  theme: lightTheme,
}

export default Tabs
