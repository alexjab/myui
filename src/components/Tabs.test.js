import React from 'react'
import renderer from 'react-test-renderer'

import defaultTheme from '../../src/core/themes'
import Tabs from '../../src/components/Tabs'

describe('<Tabs />', () => {
  it('Select a tab', () => {
    const tree = renderer.create(
      <Tabs theme={defaultTheme}>
        <Tabs.Tab>#1 Tab</Tabs.Tab>
        <Tabs.Tab>#2 Tab</Tabs.Tab>
      </Tabs>
    )

    expect(tree).toMatchSnapshot()

    const secondTab = tree.root.findAllByType(Tabs.Tab)[1]
    secondTab.props.selectTab(1)

    expect(tree).toMatchSnapshot()
  })
})
