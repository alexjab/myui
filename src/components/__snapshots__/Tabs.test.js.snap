// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tabs /> Select a tab 1`] = `
.c0 {
  border-bottom: 1px solid #dcdcdc;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c1 {
  padding: 10px;
  border-bottom: 3px solid #1a73e8;
  font-weight: 600;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c2 {
  padding: 10px;
  color: #a9a9a9;
  border-bottom: none;
  font-weight: 600;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

<div
  className="c0"
>
  <div
    className="c1"
    onClick={[Function]}
  >
    #1 Tab
  </div>
  <div
    className="c2"
    onClick={[Function]}
  >
    #2 Tab
  </div>
</div>
`;

exports[`<Tabs /> Select a tab 2`] = `
.c0 {
  border-bottom: 1px solid #dcdcdc;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c2 {
  padding: 10px;
  border-bottom: 3px solid #1a73e8;
  font-weight: 600;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c1 {
  padding: 10px;
  color: #a9a9a9;
  border-bottom: none;
  font-weight: 600;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

<div
  className="c0"
>
  <div
    className="c1"
    onClick={[Function]}
  >
    #1 Tab
  </div>
  <div
    className="c2"
    onClick={[Function]}
  >
    #2 Tab
  </div>
</div>
`;
